### ISSUE TITLE
[P2][Performance] on_devices 先等待新 batch 才交付已就緒資料，TPU 預取仍阻塞主迴圈

### ISSUE THREAD

**Issue by HHim8826** [HHim8826/Mortal#47]
審查版本：`train-parquet` 的 `76f67bfa0ad9819389cb3ce81c63b2a67f013ed6`。這是本次 shared-memory／H2D 預取改動中的性能問題，也直接對應目前約 40% 的資料等待。

[run.py:164–169](https://github.com/HHim8826/Mortal/blob/76f67bfa0ad9819389cb3ce81c63b2a67f013ed6/mortal/tpu/run.py#L164-L169) 的 `for batch in batches` 仍在訓練主執行緒呼叫 DataLoader 的 `next()`；只有 device_put 在 pool 內。每次都先取得一個新的 host batch、submit，才交出 pending 裡最舊的裝置 batch。因此，後面的 batch 一旦解碼／IPC 較慢，已經完成 H2D 的前面數批也不能交給 `step()`，預取佇列無法在資料到達時間不均勻時發揮緩衝作用，取資料與 dispatch 的主機工作也仍然串行。

### 最小重現

在 Kaggle 上直接呼叫目前原始 `on_devices()`，使用 CPU backend、小型真實 Slots／collate／device_put：上游立即提供前 5 批，在第 6 批前延遲 250 ms。

```text
取出第 1 批後，pending 中已完成的 Futures = 4
確認這 4 個 Future 都已 result() 成功
再次 next(on_devices)：第 2 批仍等待 250.33 ms
等待的其實是第 6 批；第 2 批早已準備完成
```

另驗證了順序及 120 批跨 slot 重用後的數值，CPU clone 防止 aliasing 的路徑正常；這裡不是聲稱 shared buffer 產生資料毀損。

### 實際 TPU 短程對照

環境：Kaggle TPU v5e-8、JAX 0.10.2、Torch 2.8.0+cpu，遠端 `run.py`／config 的 SHA256 與審查 commit 一致。使用實際 corpus、GRP、24 workers × 3 Rayon threads、BF16 slots、192×40、batch 1024、`--remat`、AdamW／EMA。兩邊都執行實際 `run.main()`，只在 review wrapper 裡加計時；對照組額外將相同的資料迭代器放到背景 feeder，透過容量 4 的 queue 交給主執行緒。原始 repo、既有 checkpoint 沒有修改。

兩邊各跑 1,600 steps，排除前 400 steps；剩下 1,200 steps 以各窗實際耗時合併，包含同步讀回 losses，沒有在測量段定期存 checkpoint：

| 交付方式 | 完成訓練 samples/s | 主迴圈記錄的 waiting for data |
| --- | ---: | ---: |
| 本 commit 的交付順序 | 72,316 | 44.2% |
| 審查原型：有上限的背景 feeder | 90,549 | 8.8% |

此次短程吞吐量增加約 **25.2%**。原流程的量測段中，主執行緒每批平均約 **4.18 ms** 花在 DataLoader `next()`，每次進入該呼叫前 pending 都已有完成的 Future；真正等待最舊 Future 的 `result()` 平均只有 **0.0094 ms**。所以不能僅根據 waiting 比例，就把剩餘瓶頸歸因為 worker 數不足或 H2D 帶寬。

這是短程對照，不是長時間訓練速度保證；in_order=false 會讓兩次樣本順序不同，當時主機另有下載程序，比例也不是 TPU 硬體 idle 的直接量測。背景 feeder 是驗證瓶頸的原型，尚未作為正式修正提交。

建議把 host batch 取得／H2D 排程與主執行緒 dispatch 真正解耦，使用有界的 ready queue；確保已有的 ready batches 能在上游下一批尚未到達時被消費。保持目前 slot 只有在 H2D 完成後才能回收的條件，並保留例外傳遞、提前停止、worker／thread 清理與記憶體上限。回歸測試可固定第 6 批的延遲，要求已就緒的第 2–5 批不被它阻擋。

**Comment by HHim8826** [comment]
Accurate. Fixed in 8b6e40f along the lines suggested: a feeder thread takes batches from the loader and submits each to the put pool, and the step loop takes finished puts off a queue bounded at `ready` (8), in order. A slot is still returned only after its batch's H2D has completed (inside the put), whether or not the batch has been taken. The DataLoader is iterated first on the calling thread, because that forks the workers and their PR_SET_PDEATHSIG watches the forking thread. Early stop or failure sets a stop flag, joins the feeder (at most one batch away), and lets in-flight puts finish; loader and put exceptions both reach the loop. Slots grow to workers x prefetch + ready + 2.

Kaggle v5e-8, 192x40, batch 1,024, `--remat`, 24x3 workers, corpus in /dev/shm, 5,000 steps: 84,000-90,600 samples/s with 7-20% waiting for data (windows after warm-up), against ~70,000 and ~40% before. That is in line with your prototype's 90,549.

Regression tests on the CPU backend (local): with batch 6 delayed 0.5 s, batches 2-5 are delivered in < 0.1 s each; order and values hold over 60 batches through 6 slots; closing early returns within 5 s with the feeder gone and the slots back; a loader failure and a put failure both raise in the loop. Against 76f67bf the delayed-batch test fails as your repro did.
